SimMuon/GEMDigitizer/src/GEMPadDigiClusterProducer.cc

```diff
@@ -66,24 +71,36 @@ void GEMPadDigiClusterProducer::buildClusters(const GEMPadDigiCollection &det_pa
           cl.push_back((*d).pad());
         }
         else {
-          if ((*d).bx() == startBX and (*d).pad() == cl.back() + 1) {
+          if ((*d).bx() == startBX and // same bunch crossing
+              (*d).pad() == cl.back() + 1 // pad difference is 1
+              and cl.size()<maxClusterSize_) { // max 8 in cluster
             cl.push_back((*d).pad());
           }
           else {
+            // put the current cluster in the proto collection
             GEMPadDigiCluster pad_cluster(cl, startBX);
-            out_clusters.insertDigi(part->id(), pad_cluster);
+            proto_clusters.emplace_back(part->id(), pad_cluster);
+
+            // start a new cluster
             cl.clear();
             cl.push_back((*d).pad());
-            nClusters++;
           }
         }
         startBX = (*d).bx();
       }
+      // put the last cluster in the proto collection
       if (pads.first != pads.second){
         GEMPadDigiCluster pad_cluster(cl, startBX);
-        out_clusters.insertDigi(part->id(), pad_cluster);
-        nClusters++;
+        proto_clusters.emplace_back(part->id(), pad_cluster);
       }
+    } // end of partition loop
+
+    // cluster selection: pick first maxClusters_ for now
+    unsigned int iCluster=0;
```

Comment:
it looks like this is a place where a range-based loop is not the way to go
unsigned loopMax=std::min(maxClusters_,proto_clusters.size());
for ( unsigned int i=0; i<loopMax; i++) out_clusters.insertDigi(...)

@dildick